packages/game-passes/contracts/GamePasses.sol

```diff
@@ -0,0 +1,1161 @@
+            revert ArrayLengthMismatch();
+        }
+
+        // Validate mint tokens and check max supply
+        for (uint256 i = 0; i < mintTokenIds.length; i++) {
+            TokenConfig storage mintConfig = tokenConfigs[mintTokenIds[i]];
+
+            if (!mintConfig.isConfigured) {
+                revert TokenNotConfigured(mintTokenIds[i]);
+            }
+
+            _checkMaxSupply(mintTokenIds[i], mintAmounts[i]);
+        }
+
+        // Burn tokens first
+        _burnBatch(burnFrom, burnTokenIds, burnAmounts);
+
+        // Then mint new tokens
+        _mintBatch(mintTo, mintTokenIds, mintAmounts, "");
+    }
+
+    /**
+     * @notice Allows users to burn their tokens and mint new ones with a valid EIP-712 signature
+     * @param from Address to burn from and mint to (must be msg.sender)
+     * @param burnId ID of the token to burn
+     * @param burnAmount Number of tokens to burn
+     * @param mintId ID of the token to mint
+     * @param mintAmount Number of tokens to mint
+     * @param deadline Timestamp after which the signature becomes invalid
+     * @param signature EIP-712 signature from authorized signer
+     * @dev Used for user-initiated token transformations
+     * @dev Contract must not be paused
+     * @dev Reverts if:
+     *      - Contract is paused
+     *      - from address doesn't match msg.sender
+     *      - Burn token is not configured
+     *      - Mint token is not configured
+     *      - Signature is invalid or expired
+     *      - Max supply would be exceeded for mint token
+     *      - Burn operation fails (insufficient balance)
+     */
+    function burnAndMint(
+        address from,
+        uint256 burnId,
+        uint256 burnAmount,
+        uint256 mintId,
+        uint256 mintAmount,
+        uint256 deadline,
+        bytes memory signature
+    ) external whenNotPaused {
+        address caller = _msgSender();
+
+        if (from != caller) {
+            revert InvalidSender(from, caller);
+        }
+
+        TokenConfig storage burnConfig = tokenConfigs[burnId];
+        if (!burnConfig.isConfigured) {
+            revert BurnMintNotConfigured(burnId);
+        }
+
+        // Check if mint token is configured and respects max supply
+        TokenConfig storage mintConfig = tokenConfigs[mintId];
+        if (!mintConfig.isConfigured) {
+            revert TokenNotConfigured(mintId);
+        }
+
+        BurnAndMintRequest memory request = BurnAndMintRequest({
+            caller: caller,
+            burnId: burnId,
+            burnAmount: burnAmount,
+            mintId: mintId,
+            mintAmount: mintAmount,
+            deadline: deadline,
+            nonce: nonces[caller]++
+        });
+
+        verifyBurnAndMintSignature(request, signature);
+
+        _checkMaxSupply(mintId, mintAmount);
+
+        // Burn first
+        _burn(caller, burnId, burnAmount);
+
+        // Then mint new token
+        _mint(caller, mintId, mintAmount, "");
```

Comment:
you don't need to check/update config.mintedPerWallet ?

Reply:
I don't think that's necessary.
If you can hold 10 copies for token A then you should be able to hold 10 copies of the burned/consumed version.